modules/MvwmBacker/MvwmBacker.1.in
.\" t
.\" @(#)@PACKAGE@-@VERSION@ @RELDATELONG@
.TH MvwmBacker 1 "@RELDATELONG@ (@VERSION@)" Mvwm "Mvwm Modules"
.UC
.SH NAME
MvwmBacker \- the mvwm background changer module
.SH SYNOPSIS
Module MvwmBacker

\fIMvwmBacker\fP can only be invoked by mvwm.
Command line invocation of the \fIMvwmBacker\fP module will not work.

.SH DESCRIPTION

The MvwmBacker module provides functionality to change the background
when changing desktops.  Any command can be executed to change the
backgrounds.  Actually, any arbitrary command can be sent to mvwm to
execute, so you could also do things such as changing window border
colors, etc.

.SH COPYRIGHTS
The MvwmBacker module is the original work of Mike Finger.

Copyright 1994, Mike Finger. The author makes no guarantees or
warranties of any kind about the use of this module.  Use this module
at your own risk.  You may freely use this module or any portion of it
for any purpose as long as the copyright is kept intact.

.SH INITIALIZATION
During initialization, \fIMvwmBacker\fP gets config info from
\fBmvwm\fP's module configuration database (see
.IR mvwm (1),
section
.BR "MODULE COMMANDS" ).
Available options are discussed in a later section.

.SH INVOCATION
MvwmBacker can be invoked by mvwm during initialization by inserting
the line

.nf
AddToFunc StartFunction I Module MvwmBacker
.fi

in the .mvwm2rc file.

MvwmBacker can be started using a 'Module MvwmBacker' command or stopped
using a 'KillModule MvwmBacker' command at any time when mvwm is running.

MvwmBacker must reside in a directory that is listed in the ModulePath
option of mvwm for it to be executed by mvwm.

.SH CONFIGURATION OPTIONS
The following options can be placed in the .mvwm2rc file

.IP "*MvwmBacker: Command \fI(Desk d, Page x y) command\fP"
Specifies the \fIcommand\fP to execute when the viewport matches the
arguments for the desk d, page x coordinate and y coordinate. Any or all of
these three numeric arguments can be replaced with an asterisk (*) to indicate
that any value matches, in this case Desk or Page parts can be skipped.

If either the \fIDesk\fP or the \fIPage\fP parts are omitted, the
command is not executed if only the desk or the page is switched.  If
neither is given, the command is executed only once when the module is
started.  This is not the same as using asterisks for the numeric
arguments:  if asterisks are used, the command is always executed when
only the desk or page changes, if the corresponding part is omitted, the
command is never executed when only the desk or page changes.

If the \fIcommand\fP is \fI-solid\fP MvwmBacker uses the next
argument as a color in the X database and sets the background to that
color without generating a system call to xsetroot (only single word
color names may be used).

If the \fIcommand\fP is \fIcolorset\fP MvwmBacker uses the background
specified in colorset \fIn\fP for the given desk.  Please refer to
the man page of the MvwmTheme module for details about colorsets.

Otherwise the command is sent to mvwm to execute.

.IP "*MvwmBacker: RetainPixmap"
Causes MvwmBacker to retain and publish the Pixmap with which the background
has been set. This works only for the
.IR -solid " or " colorset
commands. This is useful for applications which want to use the root
Pixmap on the background to simulate transparency (for example,
Eterm and Aterm use this method). This option should also be used for the
RootTransparent colorset option (see the MvwmTheme man page).
Note: with a colorset background this command may add a lot of memory
to the X server. For example, this adds the pixmap width times height bytes
with a TiledPixmap
image, screen_width times screen_height bytes
with a Pixmap image or a C,B,D,R,S or Y
Gradient and screen_width bytes with
a VGradient or screen height bytes with an HGradient.

.IP "*MvwmBacker: DoNotRetainPixmap"
Cancels the effect of the previous option. This is the default.

.SH RUN-TIME CONFIGURATION
It it possible to replace MvwmBacker's configuration at run-time,
although it is not yet possible to remove existing configuration
lines.  This is done by simply removing the old configuration from
withing mvwm and then read a new one.  This can be done in many
ways, for example by using an mvwm function or one of the modules
.BR MvwmCommand " or " MvwmConsole .

Example:

.nf
.sp
DestroyModuleConfig MvwmBacker*
*MvwmBacker: Command (Desk 0) -solid black
*MvwmBacker: Command (Desk 1) -solid blue
.sp
.fi

.SH OLD-STYLE OPTIONS
There is continued support for the now deprecated option:

.IP "*MvwmBacker: Desk \fId command\fP"

It is functionally equivalent to omitting the page coordinates with
\fI*MvwmBacker: Command\fP:

.nf
.sp
*MvwmBacker: Command (Desk Id) command
.sp
.fi

.SH SAMPLE CONFIGURATION
The following are excerpts from an .mvwm2rc file which describe
MvwmBacker initialization commands:

.nf
.sp
####
# Set Up Backgrounds for different desktop pages (2 desks, 3x2 pages).
####
*MvwmBacker: Command (Page 2 *) -solid steelblue
*MvwmBacker: Command (Desk 0, Page 0 0) Exec mvwm-root $[HOME]/bg2.xpm
*MvwmBacker: Command (Desk 0, Page 0 1) -solid midnightblue
*MvwmBacker: Command (Desk 0, Page 1 *) -solid yellow
*MvwmBacker: Command (Desk 1, Page * 0) -solid navy
*MvwmBacker: Command (Desk 1, Page * 1) Colorset 5
.sp
.fi

.SH AUTHOR
Mike Finger (mfinger@mermaid.micro.umn.edu)
            (Mike_Finger@atk.com)
            (doodman on IRC, check the #linux channel)
.SH Modified by
Andrew Davison (davison@cs.monash.edu.au)

Michael Han (mikehan@best.com)

Mikhael Goikhman (migo@homemail.com)


NEWS
Note, the changes for the last STABLE release start with release
2.6.0.

-------------------------------------------------------------------
Changes in release 2.6.6 (not released yet)

* Released features:

  - FvwmForm supports separator lines.
  - Support for Russian from Ivan Gayevskiy.
  - EnvMatch supports infostore variables.
  - The option "forget" to the Maximize command allows to
    unmaximize a window without changing its size and position.
  - Windows shaded to a corner (NW, NE, SW, SE) are reduced to a
    small square.
  - New option "!raise" to the WarpToWindow command.

  - The option "forget" to the Maximize command allows to
    unmaximize a window without changing its size and position.

* Bug fixes:

  - Provide a wrapper for the deprecation of XKeycodeToKeysym and
    use XkbKeycodeToKeysym() where appropriate.
  - fvwm-desktop-menu is re-written and provides better support of
    the XDG menu specification.
  - FvwmIconMan had problems displaying the hilight colour on some
    systems (64 bit issue?).
  - Globally active windows cannot take the focus if the style
    forbids programs to take foxus themselves (stlye
    !FPFocusByProgram).
  - Windows no longer jump from one position to the other which
    could happen in some cases with SnapAttraction.  Windows now
    snap to the closest window (or screen edge).
  - Removing bindings had several strange side effects that are
    fixed now (removing too many bindings; old bindings showing up
    again after another is removed; possibly other effects).
  - Windows sometimes did not get expose events (i.e. did not
    redraw properly) if they were uncovered by moving a window
    above them.  This has been fixed.

  - FvwmForm now honours a Xinerama screen specifier in its
    geometry.
  - FvwmIconMan had problems displaying the hilight colour on some
    systems (64 bit issue?).
  - Several possible format string attacks have been fixed.

-------------------------------------------------------------------
Changes in release 2.6.5 (20-Apr-2012)

* New features:

  - FvwmPager now wraps window names if SmallFont is set.

* Bug fixes:

  - Made signals registered with FVWM unblocking.
  - The "UnderMouse" option to PositionPlacement now honors the EWMH
    working area by default, if it's in use.
  - FvwmButtons handles the deleted window correctly for sub-windows.

-------------------------------------------------------------------
Changes in stable release 2.6.4 (01-Feb-2012)

* New features:

  - FvwmIconMan no longer allows for itself to be transient by using:

    Module FvwmIconMan Transient

    This is too ambiguous with valid module alias names, hence a
    transient FvwmIconMan must be created with "-Transient" as in:

    Module FvwmIconMan -Transient
  - New command InfoStore -- to store key/value pairs of information
    to relieve the burden of the SetEnv command.
  - Speed-up improvements for fvwm-menu-desktop when generating XDG
    menus by removing unnecessary stat(2) calls.
  - BusyCursor and CursorStyle have been set to make the cursor look
    like a dot during Read/PipeRead commands, by default.

* Bug fixes:

  - FvwmRearrange now understands the "ewmhiwa" option when placing
    windows.
  - Client gravity for subwindows is honoured for reparenting
    windows, such as with XEmbed, when changing the parent's
    geometry.
  - Conditional command processing of !Layer n, as in:

        Next (Xteddy, !Layer 4) Echo cuddles

    has now been fixed.
  - Fixed handling of swallowed windows for transient FvwmButtons'.
  - fvwm-menu-desktop now looks in /usr/share/applications for KDE
    legacy mode.
  - fvwm-config no longer accepts "--is-stable", "--is-released",
    "--is-final".

-------------------------------------------------------------------
Changes in stable release 2.6.3 (30-Sep-2011)

* New features:

  - New Style commands:

        TitleFormat
        IconTitleFormat

        Used to specify a window's visible name.  These new options
        deprecate the older styles:

        IndexedWindowName
        IndexedIconName
        ExactWindowName
        ExactIconName

  - New Resize argument to direction:  "automatic" which works out,
    automatically, which direction a window should be resized in
    based on which quadrant the pointer is in on the window -- this
    is best observed by the grid formed from "ResizeOutline", as the
    shape of this grid is used when calculating this.

    Furthermore, the existing option of "warptoborder" can be used in
    conjunction with "direction automatic" to warp the pointer to the
    appropriate window edge.

  - Couple of changes to the Move command:
        - The EWMH working area is now honoured by default.
        - To move a window ignoring the working area, the option
          "ewmhiwa" can be used, similar to how the Maximize
          command works.

  - New expansion placeholder $[pointer.screen] to return the screen number
    the pointer is on.

* Bug fixes:

    - A few minor ones; nothing user-visible per se.

-------------------------------------------------------------------
Changes in stable release 2.6.2 (06-Aug-2011)

* New features:

  - New MenuStyle "UniqueHotkeyActivatesImmediate" and its
    negation "!UniqueHotkeyActivatesImmediate" to retain the menu
    until the user presses one of the menu keybindings to enact
    that change in the case where there's only one item defined in
    the menu at the specified hotkey.

* Bug fixes:

  - Fix infinite loop crash when FvwmButtons quits/restarts with
    RootTransparent colorsets.
  - Fix setting window states across FVWM restarts where those
    windows do not set WM_COMMAND.
  - Windows are now correctly restored when coming out of
    fullscreen mode.

-------------------------------------------------------------------
Changes in stable release 2.6.1 (16-Apr-2011)

* New features:

  None.

* Bug fixes:

  - Make the version and check of Fribidi >= 0.19.2

modules/MvwmBacker/MvwmBacker.c
/* -*-c-*- */
/* MvwmBacker Module for mvwm.
 *
 *  Copyright 1994,  Mike Finger (mfinger@mermaid.micro.umn.edu or
 *                               Mike_Finger@atk.com)
 *
 * The author makes not guarantees or warantees, either express or
 * implied.  Feel free to use any contained here for any purpose, as long
 * and this and any other applicible copyrights are kept intact.

 * The functions in this source file that are based on part of the MvwmIdent
 * module for mvwm are noted by a small copyright atop that function, all others
 * are copyrighted by Mike Finger.  For those functions modified/used, here is
 *  the full, origonal copyright:
 *
 * Copyright 1994, Robert Nation and Nobutaka Suzuki.
 * No guarantees or warantees or anything
 * are provided or implied in any way whatsoever. Use this program at your
 * own risk. Permission to use this program for any purpose is given,
 * as long as the copyright is kept intact. */

/* Modified to directly manipulate the X server if a solid color
 * background is requested. To use this, use "-solid <color_name>"
 * as the command to be executed.
 *
 * A. Davison
 * Septmber 1994.
 */
/* This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "config.h"

#include <stdio.h>
#include <signal.h>
#include <unistd.h>
#include <sys/wait.h>

#if HAVE_SYS_BSDTYPES_H
#include <sys/bsdtypes.h> /* Saul */
#endif /* Saul */

#include <X11/Xlib.h>
#include <X11/Intrinsic.h>
#include <X11/Xatom.h>

#include "libs/FShape.h"
#include "libs/Module.h"
#include "libs/Colorset.h"
#include "libs/Parse.h"
#include "libs/PictureBase.h"
#include "libs/FRenderInit.h"
#include "libs/Graphics.h"
#include "libs/XError.h"
#include "MvwmBacker.h"

/* migo (22-Nov-1999): Temporarily until mvwm_msg is moved to libs */
#define ERR
#define mvwm_msg(t,l,f) fprintf(stderr, "[mvwm][MvwmBacker]: <<ERROR>> %s\n", f)

unsigned long BackerGetColor(char *color);

#define EXEC_CHANGED_PAGE  0x01
#define EXEC_CHANGED_DESK  0x02
#define EXEC_ALWAYS        0x04

struct Command
{
	int desk;
	struct
	{
		unsigned do_ignore_desk : 1;
		unsigned do_ignore_page : 1;
		unsigned do_match_any_desk : 1;
	} flags;
	int x;        /* Page X coordinate; -1 for glob */
	int y;        /* Page Y coordinate; -1 for glob */
	/* The command type:
	 * -1 = no command
	 *  0 = command to be spawned
	 *  1 = use a solid color background
	 *  2 = use a colorset background
	 */
	int type;
	char* cmdStr; /* The command string (Type 0)   */
	unsigned long solidColor; /* A solid color after X parsing (Type 1) */
	int colorset; /* The colorset to be used (Type 2) */
	struct Command *next;
};

typedef struct Command Command;

typedef struct
{
	Command *first;
	Command *last;
} CommandChain;

CommandChain *commands;

int current_desk = 0;
int current_x = 0;
int current_y = 0;
int current_colorset = -1;  /* the last matched command colorset or -1 */

int mvwm_fd[2];

char *Module;        /* i.e. "MvwmBacker" */
char *configPrefix;  /* i.e. "*MvwmBacker" */


/* X Display information. */

Display*        dpy;
Window          root;
int                     screen;
int MyDisplayHeight;
int MyDisplayWidth;
char* displayName = NULL;

Bool RetainPixmap = False;
Atom XA_XSETROOT_ID = None;
Atom XA_ESETROOT_PMAP_ID = None;
Atom XA_XROOTPMAP_ID = None;

unsigned char *XsetrootData = NULL;

int
ErrorHandler(Display *d, XErrorEvent *event)
{
	/* some errors are OK=ish */
	if (0 && event->error_code == BadValue)
		return 0;

	PrintXErrorAndCoredump(d, event, Module);
	return 0;
}

int main(int argc, char **argv)
{
	char *temp, *s;

	commands = xmalloc(sizeof(CommandChain));
	commands->first = commands->last = NULL;

	/* Save the program name for error messages and config parsing */
	temp = argv[0];
	s = strrchr(argv[0], '/');
	if (s != NULL)
	{
		temp = s + 1;
	}

	Module = temp;
	configPrefix = CatString2("*", Module);

	if ((argc != 6) && (argc != 7))
	{
		fprintf(stderr,
			"%s Version %s should only be executed by mvwm!\n",
			Module, VERSION);
		exit(1);
	}

	mvwm_fd[0] = atoi(argv[1]);
	mvwm_fd[1] = atoi(argv[2]);

	/* Grab the X display information now. */

	dpy = XOpenDisplay(displayName);
	if (!dpy)
	{
		fprintf(stderr, "%s:  unable to open display '%s'\n",
			Module, XDisplayName (displayName));
		exit (2);
	}
	screen = DefaultScreen(dpy);
	root = RootWindow(dpy, screen);
	MyDisplayHeight = DisplayHeight(dpy, screen);
	MyDisplayWidth = DisplayWidth(dpy, screen);
	XSetErrorHandler(ErrorHandler);
	flib_init_graphics(dpy);

	XA_XSETROOT_ID = XInternAtom(dpy, "_XSETROOT_ID", False);
	XA_ESETROOT_PMAP_ID = XInternAtom(dpy, "ESETROOT_PMAP_ID", False);
	XA_XROOTPMAP_ID = XInternAtom(dpy, "_XROOTPMAP_ID", False);

	signal (SIGPIPE, DeadPipe);

	/* Parse the config file */
	ParseConfig();

	SetMessageMask(mvwm_fd,
		       M_NEW_PAGE | M_NEW_DESK | M_CONFIG_INFO |
		       M_END_CONFIG_INFO | M_SENDCONFIG);

	/*
	** we really only want the current desk/page, and window list sends it
	*/
	SendInfo(mvwm_fd, "Send_WindowList", 0);

	/* tell mvwm we're running */
	SendFinishedStartupNotification(mvwm_fd);

	/* Recieve all messages from mvwm */
	EndLessLoop();

	/* Should never get here! */
	return 1;
}

/*
  EndLessLoop - Read until we get killed, blocking when can't read
*/
void EndLessLoop(void)
{
	while(1)
	{
		ReadMvwmPipe();
	}
}

/*
  ReadMvwmPipe - Read a single message from the pipe from mvwm
*/
void ReadMvwmPipe(void)
{
	MvwmPacket* packet = ReadMvwmPacket(mvwm_fd[1]);
	if ( packet == NULL )
	{
		exit(0);
	}
	else
	{
		ProcessMessage( packet->type, packet->body );
	}
}

void DeleteRootAtoms(Display *dpy2, Window root2)
{
	Atom type;
	int format;
	unsigned long length, after;
	unsigned char *data;
	Bool e_deleted = False;

	XA_XSETROOT_ID = XInternAtom(dpy2, "_XSETROOT_ID", False);
	XA_ESETROOT_PMAP_ID = XInternAtom(dpy2, "ESETROOT_PMAP_ID", False);
	XA_XROOTPMAP_ID = XInternAtom(dpy2, "_XROOTPMAP_ID", False);

	if (XGetWindowProperty(
		    dpy2, root2, XA_XSETROOT_ID, 0L, 1L, True,
		    XA_PIXMAP, &type, &format, &length, &after, &data) ==
	    Success && type == XA_PIXMAP && format == 32 && length == 1 &&
	    after == 0 && (Pixmap)(*(long *)data) != None)
	{
		XKillClient(dpy2, *((Pixmap *)data));
	}
	if (XGetWindowProperty(
		    dpy2, root2, XA_ESETROOT_PMAP_ID, 0L, 1L, True,
		    XA_PIXMAP, &type, &format, &length, &after, &data) ==
	    Success && type == XA_PIXMAP && format == 32 && length == 1 &&
	    after == 0 && (Pixmap)(*(Pixmap *)data) != None)
	{
		e_deleted = True;
		XKillClient(dpy2, *((Pixmap *)data));
	}
	if (e_deleted)
	{
		XDeleteProperty(dpy2, root2, XA_XROOTPMAP_ID);
	}
}

void SetRootAtoms(Display *dpy2, Window root2, Pixmap RootPix)
{
	long l_root_pix;

	l_root_pix = RootPix;
	if (RetainPixmap && RootPix != None)
	{
		XChangeProperty(
			dpy2, root2, XA_ESETROOT_PMAP_ID, XA_PIXMAP, 32,
			PropModeReplace,
			(unsigned char *) &l_root_pix, 1);
		XChangeProperty(
			dpy2, root2, XA_XROOTPMAP_ID, XA_PIXMAP, 32,
			PropModeReplace,
			(unsigned char *) &l_root_pix, 1);
	}
	else
	{
		XChangeProperty(
			dpy2, root2, XA_XSETROOT_ID, XA_PIXMAP, 32,
			PropModeReplace, (unsigned char *) &l_root_pix, 1);
	}
}

void SetDeskPageBackground(const Command *c)
{
	Display *dpy2 = NULL;
	Window root2 = None;
	int screen2;
	Pixmap pix = None;

	current_colorset = -1;

	/* MvwmBacker bg preperation */
	switch (c->type)
	{
	case 1: /* solid colors */
	case 2: /* colorset */
		dpy2 =  XOpenDisplay(displayName);
		if (!dpy2)
		{
			mvwm_msg(ERR, "MvwmBacker",
				 "Fail to create a forking dpy, Exit!");
			exit(2);
		}
		screen2 = DefaultScreen(dpy2);
		root2 = RootWindow(dpy2, screen2);
		if (RetainPixmap)
		{
			XSetCloseDownMode(dpy2, RetainPermanent);
		}
		XGrabServer(dpy2);
		DeleteRootAtoms(dpy2, root2);
		switch (c->type)
		{
		case 2:
			current_colorset = c->colorset;
			/* Process a colorset */
			if (CSET_IS_TRANSPARENT(c->colorset))
			{
				mvwm_msg(ERR,"MvwmBacker", "You cannot "
					 "use a transparent colorset as "
					 "background!");
				XUngrabServer(dpy2);
				XCloseDisplay(dpy2);
				return;
			}
			else if (Pdepth != DefaultDepth(dpy2, screen2))
			{
				mvwm_msg(ERR,"MvwmBacker", "You cannot "
					 "use a colorset background if\n"
					 "the mvwm depth is not equal "
					 "to the root depth!");
				XUngrabServer(dpy2);
				XCloseDisplay(dpy2);
				return;
			}
			else if (RetainPixmap)
			{
				pix = CreateBackgroundPixmap(
					dpy2, root2, MyDisplayWidth,
					MyDisplayHeight,
					&Colorset[c->colorset],
					DefaultDepth(dpy2, screen2),
					DefaultGC(dpy2, screen2), False);
				if (pix != None)
				{
					XSetWindowBackgroundPixmap(
						dpy2, root2, pix);
					XClearWindow(dpy2, root2);
				}
			}
			else
			{
				SetWindowBackground(
					dpy2, root2, MyDisplayWidth,
					MyDisplayHeight,
					&Colorset[c->colorset],
					DefaultDepth(dpy2, screen2),
					DefaultGC(dpy2, screen2), True);
			}
			break;
		case 1: /* Process a solid color request */
			if (RetainPixmap)
			{
				GC gc;
				XGCValues xgcv;

				xgcv.foreground = c->solidColor;
				gc = mvwmlib_XCreateGC(
					dpy2, root2, GCForeground,
					&xgcv);
				pix = XCreatePixmap(
					dpy2, root2, 1, 1,
					DefaultDepth(dpy2, screen2));
				XFillRectangle(
					dpy2, pix, gc, 0, 0, 1, 1);
				XFreeGC(dpy2, gc);
			}
			XSetWindowBackground(dpy2, root2, c->solidColor);
			XClearWindow(dpy2, root2);
			break;
		}
		SetRootAtoms(dpy2, root2, pix);
		XUngrabServer(dpy2);
		XCloseDisplay(dpy2); /* this XSync, Ungrab, ...etc */
		break;
	case -1:
	case 0:
	default:
		if(c->cmdStr != NULL)
		{
			SendMvwmPipe(mvwm_fd, c->cmdStr, (unsigned long)0);
		}
		break;
	}
}

/*
 * migo (23-Nov-1999): Maybe execute only first (or last?) matching command?
 * migo (03-Apr-2001): Yes, execute only the last matching command.
 */
void ExecuteMatchingCommands(int colorset, int changed)
{
	const Command *matching_command = NULL;
	const Command *command;

	if (!changed)
	{
		return;
	}
	for (command = commands->first; command; command = command->next)
	{
		if (!(changed & EXEC_ALWAYS) &&
		    (command->flags.do_ignore_desk ||
		     !(changed & EXEC_CHANGED_DESK)) &&
		    (command->flags.do_ignore_page ||
		     !(changed & EXEC_CHANGED_PAGE)))
		{
			continue;
		}
		if ((command->flags.do_match_any_desk ||
		     command->desk == current_desk) &&
		    (command->x < 0 || command->x == current_x) &&
		    (command->y < 0 || command->y == current_y) &&
		    (colorset < 0 || (colorset == current_colorset &&
				      colorset == command->colorset)))
		{
			matching_command = command;
		}
	}
	if (matching_command)
	{
		SetDeskPageBackground(matching_command);
	}
}

/*
  ProcessMessage - Process the message coming from mvwm
*/
void ProcessMessage(unsigned long type, unsigned long *body)
{
	char *tline;
	int change = 0;
	static Bool is_initial = True;

	switch (type)
	{
	case M_CONFIG_INFO:
		tline = (char*)&(body[3]);
		ExecuteMatchingCommands(ParseConfigLine(tline), EXEC_ALWAYS);
		break;

	case M_NEW_DESK:
		if (is_initial)
		{
			change = EXEC_ALWAYS;
			is_initial = False;
		}
		else if (current_desk != body[0])
		{
			current_desk = body[0];
			change |= EXEC_CHANGED_DESK;
		}
		ExecuteMatchingCommands(-1, change);
		break;

	case M_NEW_PAGE:
		if (is_initial)
		{
			change = EXEC_ALWAYS;
			is_initial = False;
		}
		if (current_desk != body[2])
		{
			current_desk = body[2];
			change |= EXEC_CHANGED_DESK;
		}
		if (current_x != body[0] / MyDisplayWidth ||
		    current_y != body[1] / MyDisplayHeight)
		{
			current_x = body[0] / MyDisplayWidth;
			current_y = body[1] / MyDisplayHeight;
			change |= EXEC_CHANGED_PAGE;
		}
		ExecuteMatchingCommands(-1, change);
		break;

	}
}

/*
  Detected a broken pipe - time to exit
*/
RETSIGTYPE DeadPipe(int nonsense)
{
	exit(1);
	SIGNAL_RETURN;
}

/*
  ParseConfigLine - Parse the configuration line mvwm to us to use
*/
int ParseConfigLine(char *line)
{
	int cpl = strlen(configPrefix);

	if (strlen(line) > 1)
	{
		if (strncasecmp(line, configPrefix, cpl) == 0)
		{
			if (strncasecmp(
				    line+cpl, "RetainPixmap", 12) == 0)
			{
				RetainPixmap = True;
			}
			else if (strncasecmp(
					 line+cpl,
					 "DoNotRetainPixmap", 17) == 0)
			{
				RetainPixmap = False;
			}
			else
			{
				AddCommand(line + cpl);
			}
		}
		else if (strncasecmp(line, "colorset", 8) == 0)
		{
			return LoadColorset(line + 8);
		}
	}
	return -1;
}

/*
  ParseConfig - Parse the configuration file mvwm to us to use
*/
void ParseConfig(void)
{
	char *line_start;
	char *tline;

	/* TA:  FIXME!  xasprintf() */
	line_start = xmalloc(strlen(Module) + 2);
	strcpy(line_start, "*");
	strcat(line_start, Module);

	InitGetConfigLine(mvwm_fd, line_start);
	GetConfigLine(mvwm_fd, &tline);

	while(tline != (char *)0)
	{
		ParseConfigLine(tline);
		GetConfigLine(mvwm_fd, &tline);
	}
	free(line_start);
}

Bool ParseNewCommand(
	char *parens, Command *this, int *do_ignore_desk, int *do_ignore_page)
{
	char *option;
	char *parens_run;

	parens_run = parens;

	while (*parens_run &&
	       (parens_run = GetNextFullOption( parens_run, &option)) != NULL)
	{
		char *name, *value;
		char *option_val;

		if (!*option)
		{
			free(option);
			break;
		}
		option_val = GetNextToken(option, &name);

		if (StrEquals(name, "Desk"))
		{
			if (GetNextToken(option_val, &value) == NULL)
			{
				mvwm_msg(ERR, "MvwmBacker",
					 "Desk option requires a value");
				return 1;
			}
			if (!StrEquals(value, "*"))
			{
				this->flags.do_match_any_desk = 0;
				this->desk = atoi(value);
			}
			*do_ignore_desk = False;
			free(value);
		}
		else if (StrEquals(name, "Page"))
		{
			if ((option_val = GetNextToken(
				     option_val,&value)) == NULL)
			{
				mvwm_msg(
					ERR, "MvwmBacker",
					"Page option requires 2 values");
				return 1;
			}
			if (!StrEquals(value, "*"))
				this->x = atoi(value);
			free(value);
			if (GetNextToken(option_val, &value) == NULL)
			{
				mvwm_msg(
					ERR, "MvwmBacker",
					"Desk option requires 2 values");
				return 1;
			}
			if (!StrEquals(value, "*"))
				this->y = atoi(value);
			*do_ignore_page = False;
			free(value);
		}
		free(name);
		free(option);
	}
	free(parens);
	return 0;
}

/*
  AddCommand - Add a command to the correct spot on the dynamic array.
*/
void AddCommand(char *line)
{
	char *token;
	Command *this;
	Bool do_ignore_desk = True;
	Bool do_ignore_page = True;

	this = xmalloc(sizeof(Command));
	this->desk = 0;
	memset(&(this->flags), 0, sizeof(this->flags));
	this->flags.do_match_any_desk = 1;
	this->x = -1;
	this->y = -1;
	this->type = -1;
	this->cmdStr = NULL;
	this->next = NULL;

	if (strncasecmp(line, "Desk", 4) == 0)
	{
		/* Old command style */

		line += 4;
		line = GetNextToken(line, &token);
		if (strcasecmp(token, "*") != 0)
		{
			this->flags.do_match_any_desk = 0;
			this->desk = atoi(token);
		}
		do_ignore_desk = False;
		free(token);
	}
	else if (strncasecmp(line, "Command", 7) == 0)
	{
		/* New command style */
		line += 7;
		while (*line && isspace(*line))
		{
			line++;
		}
		if (*line == '(')
		{
			int error;
			char *parens;

			line++;
			line = GetQuotedString(
				line, &parens, ")", NULL, NULL, NULL);
			if (line == NULL)
			{
				mvwm_msg(
					ERR, "MvwmBacker",
					"Syntax error: no closing brace");
				free(this);
				return;
			}
			error = ParseNewCommand(
				parens, this, &do_ignore_desk,
				&do_ignore_page);
			if (error)
			{
				free(this);
				return;
			}
			line++;
		}
	}
	else
	{
		mvwm_msg(
			ERR, "MvwmBacker",
			CatString2("Unknown directive: ", line));
		return;
	}
	this->flags.do_ignore_desk = do_ignore_desk;
	this->flags.do_ignore_page = do_ignore_page;

	/* Now check the type of command... */

	while (*line && isspace(*line))
	{
		line++;
	}
	if (strncasecmp(line, "-solid", 6) == 0)
	{
		/* Solid color command */

		line += 6;
		line = GetNextToken(line, &token);
		this->type = 1;
		/* use the root default color map */
		this->solidColor = (!token || !*token) ?
			BlackPixel(dpy, screen) :
			BackerGetColor(token);
		free(token);
	}
	else if (strncasecmp(line, "colorset", 8) == 0)
	{
		/* Colorset command */

		if (sscanf(line + 8, "%d", &this->colorset) < 1)
		{
			this->colorset = 0;
		}
		AllocColorset(this->colorset);
		this->type = 2;
	}
	else
	{
		/* Plain mvwm command */

		this->type = 0;
		this->cmdStr = xmalloc(strlen(line) + 1);
		strcpy(this->cmdStr, line);
	}

	if (commands->first == NULL)
	{
		commands->first = this;
	}
	if (commands->last != NULL)
	{
		commands->last->next = this;
	}
	commands->last = this;
}
